### ISSUE TITLE
Memory leak when multiple field is set to true in `ack`

### ISSUE THREAD

**Issue by ZiWee** [pika/pika#1294]
https://github.com/pika/pika/blob/a3ae3e5aae3fd546e9ae73bff0175f5d61a9d39a/examples/asynchronous_publisher_example.py#L263

better like:
```
    def on_delivery_confirmation(self, method_frame):
        confirmation_type = method_frame.method.NAME.split('.')[1].lower()
        ack_multiple = method_frame.method.multiple
        delivery_tag = method_frame.method.delivery_tag

        if ack_multiple:
            # NOTE: this is O(n)... might have performance issue
            # would be better if we use something like TreeMap in Java
            for k in self._deliveries.keys():
                if k <= delivery_tag:
                    item = self._deliveries.pop(k, None)
                    if item and isinstance(item, tuple) and len(item) > 2:
                        body, routing_key, send_count = item
                        if confirmation_type == 'nack':
                            if send_count < 1:
                                self._send(body, routing_key, send_count + 1)
                            else:
                                if self._send_callback:
                                    self._send_callback.on_fail(self.__get_routing_key(routing_key), body)
                        else:
                            if self._send_callback:
                                self._send_callback.on_success(self.__get_routing_key(routing_key), body)
        else:
            # ....
```

**Comment by michaelklishin** [comment]
@ZiWee can you please provide a way to reproduce and perhaps clarify what specifically you mean by the "better like" comment above?

**Comment by ZiWee** [comment]
as example  in https://www.rabbitmq.com/confirms.html#consumer-acks-multiple-parameter (this is different with AMQP)
all tags from 5 to 7 would not be remove from `self._deliveries`

you will reproduce by publishing messages by `pika/examples/asynchronous_publisher_example.py` high frequently

**Comment by lukebakken** [comment]
Sorry, I was unsubscribed from this repository and have missed new issues from the past year. I will investigate this before the next release, thank you!

**Comment by lukebakken** [comment]
This will be fixed by #1369
